test/metabase/query_processor/middleware/limit_test.clj

```diff
@@ -60,3 +60,35 @@
                             ::limit/original-limit nil}}
              (#'limit/add-default-limit query))
           "Preprocessed query should have :limit added to it"))))
+
+(deftest ^:parallel download-row-max-results-test
+  (testing "Apply `absolute-max-results` when `download-row-limit` is not set."
+    (let [query {:type  :query
+                 :query {}
+                 :info  {:context :csv-download}}]
+      (is (= {:type  :query
+              :query {:limit                 qp.i/absolute-max-results
+                      ::limit/original-limit nil}
+              :info  {:context :csv-download}}
+              (limit/add-default-limit query))))))
+
+(deftest download-row-limit-test
+  (testing "Apply custom download row limits when"
+    (doseq [[limit expected context] [[1000 1000 :csv-download]
+                                      [1000 1000 :json-download]
+                                      [1000 1000 :xlsx-download]
+                                      [1100000 1100000 :csv-download]
+                                      [1100000 1100000 :json-download]
+                                      [1100000 qp.i/absolute-max-results :xlsx-download]]]
```

Comment:
I think there's missing test coverage for these cases:
1. if `download-row-limit` is unset, `(mbql.u/query->max-rows-limit query)` returns a non-nil value above `absolute-max-results`, and context is not `:xlsx-download`, then we should apply the `(mbql.u/query->max-rows-limit query)` limit
2. `download-row-limit` is unset, `(mbql.u/query->max-rows-limit query)` returns a non-nil value, and context is `:xlsx-download`, then we should apply `absolute-max-results` as the limit

1 is arguably not essential but 2 is justified because this was the bug that I described here: https://github.com/metabase/metabase/pull/46401#discussion_r1705946254